### PR TITLE
tg-blas inner product

### PR DESCRIPTION
**Pull request by ZuseZ4**

### CODE REVIEW

File: enzyme/tools/enzyme-tblgen/caching.cpp [hunk 1]
```diff
@@ -27,10 +27,51 @@ os << "  bool need_" << name << " = false;\n";
...
       os 
 << ";\n";
-    }
-    os 
+  }
+  os 
 << "  bool cache_" << name << " = cacheMode && overwritten_" << name << " && need_" << name << ";\n";
 }
+  
+std::string get_input_mat(const llvm::DagInit *ruleDag) {
```
Comment (tgymnich) [inline enzyme/tools/enzyme-tblgen/caching.cpp:35]: nit: we are `using namespace llvm` in this file, so lets not use the llvm namespace here for consistency.

```suggestion
std::string get_input_mat(const DagInit *ruleDag) {
```

File: enzyme/tools/enzyme-tblgen/caching.cpp [hunk 2]
```diff
@@ -27,10 +27,51 @@ os << "  bool need_" << name << " = false;\n";
...
+  // if the corresponding rule is active
+  auto rules = pattern.getRules();
+  const auto nameVec = pattern.getArgNames();
+  const auto activeArgs = pattern.getActiveArgs();
+  assert(rules.size() == activeArgs.size());
+  for (size_t i = 0; i < rules.size(); i++) {
+    auto rule = rules[i];
+    const auto activeArg = activeArgs[i];
+    const auto name = nameVec[activeArg];
+    const llvm::DagInit *ruleDag = rule.getRuleDag();
```
Comment (tgymnich) [inline enzyme/tools/enzyme-tblgen/caching.cpp:61]:
```suggestion
    const DagInit *ruleDag = rule.getRuleDag();
```

File: enzyme/test/Enzyme/ReverseMode/blas/gemm_beta_f_c.ll [hunk 3]
```diff
@@ -0,0 +1,236 @@
...
+  store i64 4, i64* %m, align 16
+  store i64 4, i64* %n, align 16
+  store i64 8, i64* %k, align 16
+  store double 1.000000e+00, double* %alpha, align 16
+  store i64 4, i64* %lda, align 16
+  store i64 8, i64* %ldb, align 16
+  ;store double 0.000000e+00, double* %beta
+  store i64 4, i64* %ldc, align 16
+  call void @dgemm_64_(i8* %transa, i8* %transb, i8* %m_p, i8* %n_p, i8* %k_p, i8* %alpha_p, i8* %A, i8* %lda_p, i8* %B, i8* %ldb_p, i8* %beta, i8* %C, i8* %ldc_p) 
+  ;call void @dgemm_64_(i8* %transa, i8* %transb, i8* %m_p, i8* %n_p, i8* %k_p, i8* %alpha_p, i8* %A, i8* %lda_p, i8* %B, i8* %ldb_p, i8* %beta_p, i8* %C, i8* %ldc_p)
```
Comment (tgymnich) [inline enzyme/test/Enzyme/ReverseMode/blas/gemm_beta_f_c.ll:37]: whats going on with the commented out lines here? Maybe add a comment explaining this or remove them.

File: enzyme/tools/enzyme-tblgen/enzyme-tblgen.cpp [hunk 4]
```diff
@@ -2011,125 +2050,141 @@ void emit_fwd_rewrite_rules(const TGPattern &pattern, raw_ostream &os) {
...
 
-void emit_deriv_fnc(const StringMap<TGPattern> &patternMap, Rule &rule,
-                    StringSet<> &handled, raw_ostream &os) {
-  const auto ruleDag = rule.getRuleDag();
-  const auto typeMap = rule.getArgTypeMap();
-  const auto opName = ruleDag->getOperator()->getAsString();
-  const auto nameMap = rule.getArgNameMap();
+void emit_deriv_blas_call(DagInit *ruleDag,
+                          const StringMap<TGPattern> &patternMap,
+                          llvm::StringSet<> &handled, llvm::raw_ostream &os) {
```
Comment (tgymnich) [inline enzyme/tools/enzyme-tblgen/enzyme-tblgen.cpp:2055]:
```suggestion
                          StringSet<> &handled, raw_ostream &os) {
```

File: enzyme/tools/enzyme-tblgen/enzyme-tblgen.cpp [hunk 5]
```diff
@@ -2011,125 +2050,141 @@ void emit_fwd_rewrite_rules(const TGPattern &pattern, raw_ostream &os) {
...
-  if (Def->isSubClassOf("b")) {
-    const auto dfnc_name = Def->getValueAsString("s");
-    if (patternMap.find(dfnc_name) == patternMap.end()) {
-      PrintFatalError("calling unknown Blas function");
-    }
-    TGPattern calledPattern = patternMap.find(dfnc_name)->getValue();
-    bool derivlv23 = calledPattern.isBLASLevel2or3();
-    DenseSet<size_t> mutableArgs = calledPattern.getMutableArgs();
+  const auto dfnc_name = Def->getValueAsString("s");
+  if (patternMap.find(dfnc_name.str()) == patternMap.end()) {
```
Comment (tgymnich) [inline enzyme/tools/enzyme-tblgen/enzyme-tblgen.cpp:2059]:
```suggestion
  if (patternMap.find(dfnc_name) == patternMap.end()) {
```

File: enzyme/tools/enzyme-tblgen/enzyme-tblgen.cpp [hunk 6]
```diff
@@ -2011,125 +2050,141 @@ void emit_fwd_rewrite_rules(const TGPattern &pattern, raw_ostream &os) {
...
+    PrintFatalError("calling unknown Blas function");
+  }
+  TGPattern calledPattern = patternMap.find(dfnc_name.str())->getValue();
```
Comment (tgymnich) [inline enzyme/tools/enzyme-tblgen/enzyme-tblgen.cpp:2062]:
```suggestion
  TGPattern calledPattern = patternMap.find(dfnc_name)->getValue();
```

File: enzyme/tools/enzyme-tblgen/enzyme-tblgen.cpp [hunk 7]
```diff
@@ -2011,125 +2050,141 @@ void emit_fwd_rewrite_rules(const TGPattern &pattern, raw_ostream &os) {
...
+            "byRef ? (Type*)PointerType::getUnqual(intType) : (Type*)intType";
+      } else if (Def->isSubClassOf("transpose")) {
+        auto argStr = Def->getValueAsString("name");
+        // transpose the given trans arg, but type stays
+        typeToAdd = (Twine("type_") + argStr).str();
+      } else if (Def->isSubClassOf("use")) {
+        // we only use tmp matrices, so mat type
+        // TODO: use actual mat type, not hardcoded A
+        // auto argStr = Def->getValueAsString("name");
+        typeToAdd = (Twine("type_A")).str();
```
Comment (tgymnich) [inline enzyme/tools/enzyme-tblgen/enzyme-tblgen.cpp:2109]:
```suggestion
        typeToAdd = "type_A";
```

File: enzyme/Enzyme/Utils.cpp [hunk 8]
```diff
@@ -668,6 +668,112 @@ void callMemcpyStridedLapack(llvm::IRBuilder<> &B, llvm::Module &M,
   B.CreateCall(fn, args, bundles);
 }
 
+llvm::CallInst *getorInsertInnerProd(
+    llvm::IRBuilder<> &B, llvm::Module &M, BlasInfo blas, IntegerType *IT,
+    Type *BlasPT, Type *BlasIT, Type *fpTy, llvm::ArrayRef<llvm::Value *> args,
+    const llvm::ArrayRef<llvm::OperandBundleDef> bundles, bool byRef) {
```
[the remark this answers is not in the document]
Comment (ZuseZ4) [reply enzyme/Enzyme/Utils.cpp:674]: thx, I will try to get used to these types.

File: enzyme/Enzyme/Utils.cpp [hunk 9]
```diff
@@ -668,6 +668,112 @@ void callMemcpyStridedLapack(llvm::IRBuilder<> &B, llvm::Module &M,
...
+    PHINode *Bidx = B3.CreatePHI(intTy, 2, "Bidx");
+    PHINode *iter = B3.CreatePHI(intTy, 2, "iteration");
+    Aidx->addIncoming(ConstantInt::get(intTy, 0), init);
+    Bidx->addIncoming(ConstantInt::get(intTy, 0), init);
+    iter->addIncoming(ConstantInt::get(intTy, 0), init);
+
+    Value *Ai = B3.CreateInBoundsGEP(BlasPT, matA, Aidx, "A.i");
+    Value *Bi = B3.CreateInBoundsGEP(BlasPT, matB, Bidx, "B.i");
+    // Value *Ai = B3.CreateInBoundsGEP(fpTy, matA, Aidx, "A.i");
+    // Value *Bi = B3.CreateInBoundsGEP(fpTy, matB, Bidx, "B.i");
```
[the remark this answers is not in the document]
Comment (ZuseZ4) [reply enzyme/Enzyme/Utils.cpp:757]: wip, I will finish it tmrw

File: enzyme/test/Enzyme/ReverseMode/blas/gemm_beta_f_c.ll (hunk 3 above)
Comment (ZuseZ4) [reply enzyme/test/Enzyme/ReverseMode/blas/gemm_beta_f_c.ll:37]: this test is not checking what it should check later, I just use it to trigger abi missmatches in my implementation of an inner product.

File: enzyme/Enzyme/Utils.cpp [hunk 10]
```diff
@@ -668,6 +668,112 @@ void callMemcpyStridedLapack(llvm::IRBuilder<> &B, llvm::Module &M,
...
+  Function *FDot =
+      cast<Function>(M.getOrInsertFunction(dot_name, FDotT).getCallee());
+
+  // now add the implementation for the inner_prod call
+  F->setLinkage(Function::LinkageTypes::InternalLinkage);
+#if LLVM_VERSION_MAJOR >= 16
+  F->setOnlyAccessesArgMemory();
+#else
+  F->addFnAttr(Attribute::ArgMemOnly);
+#endif
```
Comment (wsmoses) [inline enzyme/Enzyme/Utils.cpp:702]: You might as well mark the whole function as read only as well, no?

Comment (ZuseZ4) [reply enzyme/Enzyme/Utils.cpp:703]: done, thx

**Comment by ZuseZ4** [comment]
We can optimize the number of matrices which we cache further, but I'll add that once the other PR for more complex rules is merged, which will allow me to support more combinations and therefore testcases.

**Review by wsmoses — APPROVED** [review]

**Pull request merged**